### PR TITLE
Iceberg registering in github actions

### PR DESCRIPTION
**Pull request by hogoww**
Hello,

This PR "fixes" #581 by redirecting users to the setup-SmalltalkCI link on which I will do another PR to add the requirements to make Iceberg work in github actions.
Basically, we need to fetch all of the commits because Iceberg always creates a full history as far as I can tell.
This is achieve with the option fetch-depth: 0 for actions/checkout@v2
```yaml
- uses: actions/checkout@v2
        with:
          fetch-depth: 0
```
(Will do that PR later today)

I also put Iceberg in a preload rather than postload, to be able to use it during the loading.
(In my case, this was necessary for resource usage).
If you think there should be another explanation on how to Iceberg in the CI, let me know which file I should add it to.

Thanks again to @theseion for the support & @fniephaus for the solution !

### CODE REVIEW

File: repository/SmalltalkCI-Core.package/SmalltalkCI.class/instance/preLoad.st [hunk 1]
```diff
@@ -0,0 +1,3 @@
+executing
+postLoad
```
Comment (fniephaus) [inline repository/SmalltalkCI-Core.package/SmalltalkCI.class/instance/preLoad.st:2]: I think this should say `preLoad` .

File: repository/SmalltalkCI-Core.package/SmalltalkCI.class/instance/postLoad.st [hunk 2]
```diff
@@ -1,4 +1,3 @@
 executing
 postLoad
```
Comment (fniephaus) [inline repository/SmalltalkCI-Core.package/SmalltalkCI.class/instance/postLoad.st:2]: `postLoad` doesn't seem to be needed anymore, so please drop it.

File: repository/SmalltalkCI-Core.package/SmalltalkCI.class/instance/load.st [hunk 3]
```diff
@@ -1,4 +1,5 @@
 executing
 load
+	self preLoad.
 	self class withBuildStatusReportingDo: [ self basicLoad ].
 	self postLoad
```
Comment (fniephaus) [inline repository/SmalltalkCI-Core.package/SmalltalkCI.class/instance/load.st:5]: `postLoad` doesn't seem to be needed anymore, so please drop it.

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st [hunk 4]
```diff
@@ -0,0 +1,10 @@
+executing
+preLoad
+	[ (IceRepositoryCreator new
+		location: projectDirectory asFileReference;
+		createRepository) register. ] 
+		on: Error 
+		do: [ :e |
+		      	 self class printWarningToStdout: 'Could not register the tested repository with Iceberg.'.
+		      	 self class printWarningToStdout: 'If you need to use Iceberg features in the CI, see github.com/hpi-swa/setup-smalltalkCI'.
```
Comment (fniephaus) [inline repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: This somewhat assumes that smalltalkCI runs on GHA, which is a common but not the only case.

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st [hunk 5]
```diff
@@ -0,0 +1,10 @@
+executing
+preLoad
+	[ (IceRepositoryCreator new
+		location: projectDirectory asFileReference;
+		createRepository) register. ] 
+		on: Error 
+		do: [ :e |
+		      	 self class printWarningToStdout: 'Could not register the tested repository with Iceberg.'.
```
Comment (fniephaus) [inline repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:8]: Can this just be a long string instead of multiple calls to `self class printWarningToStdout`?

**Review by fniephaus — CHANGES_REQUESTED** [review]

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st (hunk 4 above)
Comment (hogoww) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: Are the yml different when not on GHA? I'm not familiar with such runs

**Comment by hogoww** [comment]
Thanks a lot for the review !
(And nice catch of my remaining issue is the non-GHA builds)

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st (hunk 4 above)
Comment (fniephaus) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st]: Yes, the main README.md has some examples: https://github.com/hpi-swa/smalltalkCI#gitlab-ci-template.

Comment (hogoww) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st]: Would you be okay if that said that is only valid for GHA ?
Don't really know whether that'd work for travis et al.

**Comment by fniephaus** [comment]
Isn't there an entry missing for preLoad in:
https://github.com/hpi-swa/smalltalkCI/blob/b58ba5b309013ccf1d8623dac82d0e4eb69d2568/repository/SmalltalkCI-Core.package/SmalltalkCI.class/methodProperties.json#L115-L138

The 32bit builds are currently failing because GHA runners have been bumped to Ubuntu 22.04. I'll take a look at this in a few days.

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st (hunk 4 above)
Comment (fniephaus) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st]: I'd add to the note that the full Git history must be available for Iceberg and then link to the setup-smalltalkCI docs.

**Comment by fniephaus** [comment]
Please address https://github.com/hpi-swa/smalltalkCI/pull/582#issuecomment-1328007708. 32-bit builds should be working again.

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+preLoad
+	[ (IceRepositoryCreator new
+		location: projectDirectory asFileReference;
+		createRepository) register. ] 
+		on: Error 
+		do: [ :e |
+		      	 self class printWarningToStdout: (String streamContents: [ :s |
+			 	s << 'Could not register the tested repository with Iceberg.';
+				  cr;
+				  << 'If you need to use Iceberg features in the CI, you need to have the full git history available.'
```
Comment (fniephaus) [inline repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:11]: This seems to be missing a `;`:
```suggestion
				  << 'If you need to use Iceberg features in the CI, you need to have the full git history available.';
```

File: repository/SmalltalkCI-Core.package/SmalltalkCI.class/methodProperties.json [hunk 7]
```diff
@@ -130,6 +130,7 @@
 		"prepareForTesting" : "fn 11/1/2017 13:47",
 		"projectDirectory" : "fn 11/9/2017 11:41",
 		"projectDirectory:" : "fn 11/9/2017 11:41",
+		"projectDirectory:" : "Hogoww 12/13/2022 19:29",
```
Comment (fniephaus) [inline repository/SmalltalkCI-Core.package/SmalltalkCI.class/methodProperties.json:133]:
```suggestion
		"preLoad" : "Hogoww 12/13/2022 19:29",
		"projectDirectory" : "fn 11/9/2017 11:41",
		"projectDirectory:" : "fn 11/9/2017 11:41",
```

File: repository/SmalltalkCI-Core.package/SmalltalkCI.class/methodProperties.json [hunk 8]
```diff
@@ -126,6 +126,7 @@
 		"isCoverageTestingEnabled" : "PeterUhnak 10/20/2017 12:33",
 		"load" : "fn 11/20/2017 11:56",
 		"performMetacelloLoad:" : "fn 4/15/2016 10:47",
+		"preLoad:" : "Hogoww 12/13/2022 19:29",
```
Comment (fniephaus) [inline repository/SmalltalkCI-Core.package/SmalltalkCI.class/methodProperties.json:129]: This is unfortunately still incorrect. `preLoad` does not take an argument, so it should read:

```suggestion
		"preLoad" : "Hogoww 12/13/2022 19:29",
```

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st [hunk 9]
```diff
@@ -0,0 +1,15 @@
+executing
+preLoad
+	[ (IceRepositoryCreator new
+		location: projectDirectory asFileReference;
+		createRepository) register. ] 
+		on: Error 
+		do: [ :e |
+		      	 self class printWarningToStdout: (String streamContents: [ :s |
+			 	s << 'Could not register the tested repository with Iceberg.';
```
Comment (fniephaus) [inline repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: Is this always printed, including when someone does not need Iceberg and the repo was only shallow cloned? If so, maybe we should silence the error here? Any thoughts, @theseion?

Comment (theseion) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: You're right. It's not bad but a bit ugly. I wonder whether we could make registration an option of the run script? E.g.,:
```bash
./run.sh -s Pharo64-10 --headful -d ../repo/.smalltalk.ston --register
```

I guess, the GitHub action would then have the equivalent option.

**Comment by hogoww** [comment]
@fniephaus Sorry about all those small mistakes.
Thanks for catching them !

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st (hunk 9 above)
Comment (theseion) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: Good point. The config file is probably a better place for the setting.

Comment (hogoww) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: That's what I'd tend to do.
However, I'd have to investigate how to do it.
Does that sounds okay @theseion ?

Comment (fniephaus) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: For the implementation, maybe take a look at https://github.com/hpi-swa/smalltalkCI/commit/0a4eb665b36f7e46c866b4e03c189330f33b384a.

Comment (theseion) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: Sounds good. I would expand the comment a bit and either point to documentation on what "full history" means or explain the necessary steps (it's clear what it means for a local repository but not for a CI system).

Comment (hogoww) [reply repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:9]: > For the implementation, maybe take a look at [0a4eb66](https://github.com/hpi-swa/smalltalkCI/commit/0a4eb665b36f7e46c866b4e03c189330f33b384a).

Thanks for the pointer, I will try to do it after my phd defense.

**Comment by hogoww** [comment]
I finally found the time to look into it.
I'm not sure whether that's appropriate to not pollute over CIs.
Please let me know.

**Comment by hogoww** [comment]
CI looks stuck here, although the action seem completed in the action tab

**Comment by hogoww** [comment]
Discard last commit comment, this is a separate issue.
This looks ready from a code standpoint, awaiting review :)

File: README.md [hunk 10]
```diff
@@ -422,6 +422,7 @@ SCIMetacelloLoadSpec {
   #passwordEnvVar : 'GITHUB_TOKEN',                   // Environment variable containing the password used for authentication
   #useLatestMetacello : true,                         // Upgrade Metacello before loading
   #version : '1.0.0'                                  // Define MC version (for MC Configurations only)
+  #usesIceberg : true				      // Pharo Only | Enable the Iceberg registering for the tested repository
```
Comment (theseion) [inline README.md:425]:
```suggestion
  #usesIceberg : true				      // Pharo Only | Register the tested repository with Iceberg
```

File: repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st [hunk 11]
```diff
@@ -0,0 +1,11 @@
+accessing
+usesIceberg: aBoolean
+	"Pharo specific option allowing the use of Iceberg Features.
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st:3]:
```suggestion
	"Pharo specific option: register the tested repository with Iceberg.
```

File: repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+	To enable this, you also need to get the full history for your repository.
+	Therefore you need to modify your yaml to have the following:
+steps: 
+	- uses: actions/checkout@v2
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st:7]: I think version 3 is current.
```suggestion
	- uses: actions/checkout@v3
```

File: repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st [hunk 13]
```diff
@@ -0,0 +1,11 @@
+accessing
+usesIceberg: aBoolean
+	"Pharo specific option allowing the use of Iceberg Features.
+	To enable this, you also need to get the full history for your repository.
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st:4]:
```suggestion
	For registration with Iceberg to work, the repository must not be shallow.
```

File: repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+	Therefore you need to modify your yaml to have the following:
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st:5]:
```suggestion
	Therefore, the GitHub workflow needs to be configured as follows:
```

File: repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+steps: 
+	- uses: actions/checkout@v2
+		with: 
+			fetch-depth: 0 #option fetching all commits.
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg..st:9]:
```suggestion
			fetch-depth: 0 # fetch all commits (for Iceberg)
```

File: repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg.st [hunk 16]
```diff
@@ -0,0 +1,11 @@
+accessing
+usesIceberg
+	"Pharo specific option allowing the use of Iceberg Features.
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg.st:3]: Make the same changes to the text here as in `#usesIceberg:`.

File: repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg.st [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+	To enable this, you also need to get the full history for your repository.
+	Therefore you need to modify your yaml to have the following:
+steps: 
+	- uses: actions/checkout@v2
+		with: 
+			fetch-depth: 0 #option fetching all commits.
+	"
+	^ usesIceberg
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SCIPharoMetacelloLoadSpec.class/instance/usesIceberg.st:11]:
```suggestion
	^ usesIceberg ifNil: [ false ]
```

File: repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st [hunk 18]
```diff
@@ -0,0 +1,20 @@
+executing
+preLoad
+	(spec loading anySatisfy: [ :aLoadingSpec |
+	      aLoadingSpec usesIceberg isNotNil
```
Comment (theseion) [inline repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st:4]: Nil check not necessary with the change I propose in `#usesIceberg`.

**Review by theseion — CHANGES_REQUESTED** [review]

**Comment by hogoww** [comment]
@theseion Thanks for all the great suggestions !
Everything is integrated.
I did one commit rather than commit all your suggestion to avoid multiple CI trigger and mostly useless commits.

**Review by theseion — APPROVED** [review]

File: README.md [hunk 19]
```diff
@@ -422,6 +422,7 @@ SCIMetacelloLoadSpec {
   #passwordEnvVar : 'GITHUB_TOKEN',                   // Environment variable containing the password used for authentication
   #useLatestMetacello : true,                         // Upgrade Metacello before loading
   #version : '1.0.0'                                  // Define MC version (for MC Configurations only)
+  #usesIceberg : true				      // Pharo Only | Register the tested repository with Iceberg
```
Comment (fniephaus) [inline README.md:425]: I'm not sure we should call this `#usesIceberg` because it is a bit vague. Should we rename this to `#registerInIceberg`, which seems to be more descriptive if I look at what the `preLoad` hook does. What do you think?

**Review by fniephaus — COMMENTED** [review]
Thanks for updating this PR!

File: README.md (hunk 19 above)
Comment (hogoww) [reply README.md:425]: I thought something similar, albeit longer than yours which I like.
This is particularly more true with all @theseion corrections.
Okay with you @theseion ?
(I have to modify the text representation of the code because Pharo wants to commit way too much stuff, so I'd rather not redo it a 4th time :D)

Comment (theseion) [reply README.md:425]: Yes, I fully agree.

**Comment by hogoww** [comment]
Done, I hope I haven't forgotten anything.
I'll have to update the associated PR in the setup.

File: README.md [hunk 20]
```diff
@@ -422,7 +422,7 @@ SCIMetacelloLoadSpec {
   #passwordEnvVar : 'GITHUB_TOKEN',                   // Environment variable containing the password used for authentication
   #useLatestMetacello : true,                         // Upgrade Metacello before loading
   #version : '1.0.0'                                  // Define MC version (for MC Configurations only)
-  #usesIceberg : true				      // Pharo Only | Register the tested repository with Iceberg
+  #registerInIceberg : true				      // Pharo Only | Register the tested repository with Iceberg
```
Comment (theseion) [inline README.md:425]:
```suggestion
  #registerInIceberg : true			// Pharo Only | Register the tested repository with Iceberg
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: README.md [hunk 21]
```diff
@@ -422,7 +422,7 @@ SCIMetacelloLoadSpec {
   #passwordEnvVar : 'GITHUB_TOKEN',                   // Environment variable containing the password used for authentication
   #useLatestMetacello : true,                         // Upgrade Metacello before loading
   #version : '1.0.0'                                  // Define MC version (for MC Configurations only)
-  #registerInIceberg : true				      // Pharo Only | Register the tested repository with Iceberg
+  #registerInIceberg : true			// Pharo Only | Register the tested repository with Iceberg
```
Comment (theseion) [inline README.md:425]: Sorry for the hassle. My last change request didn't quite show what I wanted it to, which is to align the comments properly. I think the issue is that there are tabs in that line, the other lines contain only spaces. Let me try again:
```suggestion
  #registerInIceberg : true                           // Pharo Only | Register the tested repository with Iceberg
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: README.md (hunk 21 above)
Comment (hogoww) [reply README.md]: No worry, you did all the work there, just have to click one button ;)

File: README.md [hunk 22]
```diff
@@ -422,6 +422,7 @@ SCIMetacelloLoadSpec {
   #passwordEnvVar : 'GITHUB_TOKEN',                   // Environment variable containing the password used for authentication
   #useLatestMetacello : true,                         // Upgrade Metacello before loading
   #version : '1.0.0'                                  // Define MC version (for MC Configurations only)
+  #registerInIceberg : true                           // Pharo Only | Register the tested repository with Iceberg
```
Comment (fniephaus) [inline README.md:425]: To be consistent, it should probably read:
```suggestion
  #registerInIceberg : true                           // Pharo Only | Register the tested repository in Iceberg
```

**Comment by fniephaus** [comment]
is this good to go now, @hogoww @theseion ?

**Comment by hogoww** [comment]
Good for me

**Review by theseion — APPROVED** [review]
LGTM

**Review by fniephaus — APPROVED** [review]
LGTM

**Comment by Rinzwind** [comment]
There seems to be a bug in the following:

https://github.com/hpi-swa/smalltalkCI/blob/f58749eb342884e7233046aacffdeb270240ca46/repository/SmalltalkCI-Pharo-Core.package/SmalltalkCIPharo9.class/instance/preLoad.st#L2-L4

The message `#registerInIceberg` is only understood by instances of SCIPharoMetacelloLoadSpec, it is not understood by instances of other classes in the SCIAbstractLoadSpec hierarchy (such as SCIPharoCustomScript).

**Comment by fniephaus** [comment]
I have [reverted the merge](https://github.com/hpi-swa/smalltalkCI/commit/44a709d5d133b7e02bbc11a50aba263abb2a813c), assuming this has caused failures on @Rinzwind's end. Please take a look, @hogoww.

**Comment by hogoww** [comment]
Of course, thank you @fniephaus !

A quick fix could be to simply add a canUnderstand:
Would that be okay?
I don't know much about the other supported systems

**Comment by fniephaus** [comment]
I'd like to avoid #canUnderstand:. Why not add a method returning false to `SCIAbstractLoadSpec`, which I believe is the super class of all load specs?

**Comment by hogoww** [comment]
> I'd like to avoid #canUnderstand:. Why not add a method returning false to `SCIAbstractLoadSpec`, which I believe is the super class of all load specs?

I wanted to avoid non Pharo components.
But if you're okay with that, that'd indeed be the easiest solution.

**Comment by fniephaus** [comment]
Yes, it's an easy and clean solution. 🙂

**Comment by hogoww** [comment]
Sure then !
